### PR TITLE
Made step indicators clickable.

### LINKED ISSUE #118: step indicators should clickable
Can you make the Step indicators clickable, so they can be used to jump to a step.  Currently, this is only possible by clicking on the step title.  However, not everyone will StepTitle

### PR DESCRIPTION
**Pull request by earshinov**
Fixes #118.

# Description of the changes

The changes basically boil down to moving step indicators from `li:after` to `li div a:before`. The diff looks so huge because there is a lot of boilerplate in the styles.

Just like the text, step indicators are now highlighted on hover when they are clickable. While text is darkened by 20%, a border of an empty circle is darkened by 10, and background of a filled color is darkened by 5. This is because a change in, for example, background color is more visible than in text color.

# Testing

I have tested my changes using [angular-archwizard-demo][1] (very well-made, thank you!) using all layouts, both in a horizontal and a vertical location.

Tested browsers:
- Google Chrome 67.0.3396.99
- Microsoft Edge Microsoft Edge 38.14393.2068.0

I have also run `npm test` which, of course, did not report any problems.

# Implementation notes

Thanks to absolute positioning you use for step indicators, the clickable area in horizontal layouts is as follows (marked with blue):

clickable area

I think, this is the desired behavior.

# Compatibility notes

The changes might break wizard styling for developers who customized step indicators styles using `li:after` selector. They should now use `li div a:before` instead.

[1]: <https://github.com/madoar/angular-archwizard-demo/>

### CODE REVIEW

**Comment by madoar** [comment]
First, thank you for your Pull Request!

I'm not sure I like the approach of replacing `:after` with `:before` in this case.
One reason being, that it requires a lot of changes in the stylesheets.

Some time ago, I've invested some thoughts into how to solve #118.
The solution I've thought of at that time was to move `[awGoToStep]="step"`, from the link element in the line `<a [awGoToStep]="step">`, to the div element, in the live `<li *ngFor="let step of wizardSteps"` of the wizard navigation bar.

This change makes the whole div element, i.e. the label and the indicator itself, usable for navigation.

What my solution doesn't include are the additional style changes necessary to make the step indicator react the same way as the label, when the user interacts with it, i.e. when the user hovers it etc.
I'm not sure yet, how much work it would take to add these changes.

What do you think?

**Comment by madoar** [comment]
> The solution I've thought of at that time was to move [awGoToStep]="step", from the link element in the line <a [awGoToStep]="step">, to the div element.
>> As far as I understand, in this case you will still have to change indicator placement from li:before to something that is inside the <div [awGoToStep]="step"> to make indicators clickable.

During my tests, this wasn't the case, i.e. the mentioned change made it possible to click on the indicator to move to another step.

**Comment by earshinov** [comment]
Could you share your experiments so that I can have a look?

**Comment by madoar** [comment]
I've made an error in my previous comment:
I've moved `[awGoToStep]="step"` from the `a` element to the `li` element, not to the `div` element!

Reproducing my experiments is quite easy:
- first, clone this repository and the demo repository at https://github.com/madoar/angular-archwizard-demo
- afterwards, make the mentioned change, i.e. move `[awGoToStep]="step"` from the `a` element to the `li` element in https://github.com/madoar/angular-archwizard/blob/master/src/components/wizard-navigation-bar.component.html
- build the cloned demo project, using your local and changed `angular-archwizard` build

Now, all in the demo shown wizards should be navigable by clicking on the indicator.

**Comment by earshinov** [comment]
Yes, moving `[awGoToStep]` to `li` makes sense.

Does not it bother you that the resulting markup will not be semantically correct? We will have a `<li>` that is actually clickable and an `<a>` inside it that does not really surves as a navigatable element (because `<li>` does). Just curious before I start experimenting.

**Comment by earshinov** [comment]
By the way, I have made a test page for experimenting with layout / styling:
https://plnkr.co/edit/SuBJXsd48c7jPlteoEIk?p=preview

**Comment by earshinov** [comment]
A note to myself: it there are changes in the styles after all, the custom css demo (https://github.com/madoar/angular-archwizard-demo/blob/master/src/app/custom-css/custom-css.component.css) will have to be adjusted accordingly.

**Comment by madoar** [comment]
Yes, you're right about the semantic.
Still I think it's best to move `[awGoToStep]` to the `li` element.
To improve the correctness of the resulting semantics, we could also think about changing the `a` element to something else, like an inner `div` or a `h3` element.

**Comment by earshinov** [comment]
@madoar , Sorry that I disappeared for so long.

I would like to draw your attention to another issue. While I was testing my changes, I noticed that `navigationSymbol.fontFamily` is applied not only to step indicators, but also to step labels: https://github.com/madoar/angular-archwizard/issues/130

I don't think that issue can be fixed if we continue to attach step indicators to `li:before`, because this way indicator styles will always be inherited by step labels. With that in mind, I think layout changes are inevitable in the end, and the best course of action is to rework the layout so that:

  - the new layout allows us to fix both this issue and the fontFamily issue;
  - the new layout is flexible and future-proof enough to prevent problems similar to those we now have with the `:before / :after` approach.

**Comment by madoar** [comment]
Yes, I agree with you.

I believe we should move the step indicator from `li:after` to a `div` container.
This has multiple advantages:
- for one, the styling of the step indicator becomes easier
- in addition, this change makes it possible to make the content of the step indicator more flexible.
   Currently it's only possible to use text as the step indicator, because the css command `attr()` only returns text.
   With a div container, we could allow additional context types for the step indicator, for example an image.

A follow up question is, do we want to continue to define the connection lines between the indicators in `li:before`, or do we also want to change this?

Do you want to give it a try?

**Comment by earshinov** [comment]
@madoar , Yes, I am ready to take on this challenge :)

For now, I suggest the following DOM structure:

```
<aw-wizard-navigation-bar class="horizontal large-empty-symbols">
  <ul class="step-indicator steps-3">
    <li class="done navigable">
      :before
      <a>
        <div class="label">Steptitle 1</div>
        <div class="step-indicator" style="font-family: FontAwesome;"></div>
      </a>
    </li>
    ...
  </ul>
</aw-wizard-navigation-bar>
```

Rationale:

1. We still need a common parent for the label and the step indicator so that both can be highlighted using CSS `:hover` when either one is hovered. In my suggestion this parent is `<a>`. The `<li>` element cannot be such parent because the `<li>` has to occupy space till the next `<li>`  and we do not want to make all this area styled on hover.

2. We can leave the current scheme when `<a>` is nested in a `<div>` (not vice versa as in my suggestion):

```
      <div class="step">
        <a>Steptitle 1</a>
        <div class="step-indicator" style="font-family: FontAwesome;"></div>
      </div>
```

There is little difference technically - what really matters is where we put the `[awGoToStep]` directive. But my suggestion seems more semantically correct.

3. As you proposed, the step indicator resides in a separate `<div>` where we can apply the configured symbol and font-family in isolation from the label.

4. Regarding the connection lines, I guess we can leave them in `li:before` for now. Moving them to separate DOM elements can be done rather easily on demand.

5. I think, the DOM structure I propose will suit our current needs in terms of layout without introducing additional wrappers.

If you agree with my suggestion, I am ready to roll out a quick prototype.

**Comment by earshinov** [comment]
Well, actually I can share the prototype right now:

https://plnkr.co/edit/DLZ71X7MxtzpfSGnnYJK?p=preview

Feel free to tinker with it.

Here is a GitHub repo with full history:

https://github.com/earshinov/angular-archwizard-layout-proto

For now, I've prepared the horisontal layout only.

**Comment by madoar** [comment]
Your proposal looks good to me.

Just some small things, I think are important:
- first, we need to ensure, that the indicator is configurable, i.e. we can't set a fixed style for the indicator.
- in addition to the previous point, it seems like the current indicator implementation doesn't work with fontawesome 5 (see #134). Therefore, we should also try to fix this problem, if we can.

About the question, whether we should wrap the `a` element inside a `div` element or the other way round:
Basically I'm fine either way. 
My "problem" with using an `a` element as a container is, that it I'm not accustomed to use an `a` element as a container.
In my experience `a` elements are used as links and should therefore contain only text or in some cases an image.
Maybe that's just me...

**Comment by earshinov** [comment]
@madoar , Sorry that I went missing again. Crunch time at work…

So, I guess I will now get down to applying the demo layout to angular-archwizard's source code.

**Comment by earshinov** [comment]
Regarding `a` elements, I understand perfectly. Wrapping block elements in `<a>` seemed peculiar to me when I first came across that. But it always worked, even in IE6 times, despite the fact that `<a>` was indeed defined as an inline element. When HTML5 came out, the inline-vs-block notion was in fact superseded by so called content categories, and `<a>` now belongs to the same "Flow content" category as for example `<p>` (https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#Flow_content). So there are no problems with using `<a>` around block elements, neither practical nor semantic.

By the way, I have just found a perfect illustration. HTML code of the demo layout actually mirrors the HTML code of this Google widget:

google widget

**Comment by earshinov** [comment]
As for #134, am I right that the problem is that currently it is impossible to customize `font-weight` in addition to `font-family` and `symbol`? If it is so, I am going to put forward a bold proposal: why don't we let the user to customize step symbols with a template, just like step titles? It could look like this:

```
<aw-wizard-step>
    <ng-template awWizardStepIcon><i class="far fa-text"></i></ng-template>
    <ng-template awWizardStepTitle><span>Performance</span></ng-template>
    ...
```

which seems much cleaner and easier to code than

```
<aw-wizard-step [navigationSymbol]="{ symbol: '&#xf0ce;', fontFamily: '&quot;Font Awesome 5 Pro&quot;' }">
    <ng-template awWizardStepTitle><span>Performance</span></ng-template>
    ...
```

To preserve backward compatibility, we can continue to support `symbol` and `fontFamily` along with the template approach.

**Comment by madoar** [comment]
Regarding #134: yes, it seems that this is the problem.
I agree to your proposal, this seems to be the best solution.
But like you've also written, we should provide multiple methods to provide the navigation symbol:
- for one we should continue to support the "old way" using `[navigationSymbol]="{ symbol: '&#xf0ce;', fontFamily: 'Font Awesome 5 Pro' }"`.
   This method is the fastest one for developers who simply want to show a single symbol or letter in the navigation symbol.
   Also I would like to extend the navigation symbol input to also support image links and maybe other content types too
- in addition, I we should support templates, where the developer can freely design his own navigation symbols.

**Comment by earshinov** [comment]
Had to open a separate PR for the new version: https://github.com/madoar/angular-archwizard/pull/143

**Comment by earshinov** [comment]
@madoar , Please take a look at

- https://github.com/madoar/angular-archwizard/pull/143
- https://github.com/madoar/angular-archwizard-demo/pull/22
- https://github.com/madoar/angular-archwizard/pull/144
